### ISSUE TITLE
Testing custom error handlers

### ISSUE THREAD

**Issue by gonzacastro** [tiangolo/fastapi#5104]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class BaseError(Exception):
    def __init__(self, message, status_code) -> None:
        self.message = message
        self.status_code = status_code

async def error_handler(request: Request, exc: BaseError):
    return JSONResponse(content={"message": exc.message}, status_code=exc.status_code)
```

### Description

Given the example code, which is only a custom exception class and its handler, what I was trying while making the tests was to cover the handler lines but it keeps being uncovered, even when I force the exception to be raised. By the time, it's not that relevant because it's only one line but it'd get a pain if one handler is larger...
I tried several things, from raising the exception myself to adding the handler to the test client, so please, if anyone know about this I'd really appreciate the help.

### Operating System

Windows

### FastAPI Version

0.78.0

### Python Version

3.9.13

**Comment by odiseo0** [comment]
Btw, this code is similar to yours

```py
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse


class BaseError(Exception):
    def __init__(self, message: str, status_code: int) -> None:
        self.message = message
        self.status_code = status_code


app = FastAPI()

@app.exception_handler(BaseError)
async def error_handler(request: Request, exc: BaseError):
    print("Hello")
    return JSONResponse(content={"message": exc.message}, status_code=exc.status_code)


@app.get("/")
def index():
    raise BaseError(message="Error", status_code=404)
```

The output I get when visiting `/` is the following:
```console
Hello
INFO:     127.0.0.1:63867 - "GET / HTTP/1.1" 404 Not Found
```

Maybe you're missing the `exception_handler` decorator

**Comment by odiseo0** [comment]
Understood. How are you writing the test and what is the behavior that you expect from the `exception_handler`?

**Comment by JarroVGIT** [comment]
> I'm patching some random app function with a BaseError side effect so it triggers the handler.

I don’t think this works the way you think it works. Especially if you say that when testing the handler directly, and you didn’t expect a coroutine to be returned by an asynchronous function🤔 your (async) test client needs to await the function for it to return a value.

Provide the testcase that you tried, including any fixtures and mocking, maybe that clarifies.

**Comment by gonzacastro** [comment]
After several tries, I could achieve to cover the file with tests.
```
@pytest.mark.anyio
async def test_error_handler():
    response = await error_handler(None, error)
    assert response.body == {'message': 'test'}
```
Thank you guys for replying!
